File: app/frontend/types/bid.ts
import { User } from './auth';

export interface Bid {
  id?: number;
  job_id: number;
  supplier_id?: number;
  amount: number;
  message: string;
  status?: 'pending' | 'accepted' | 'rejected' | 'withdrawn';
  created_at?: string;
  updated_at?: string;
  supplier?: User;
  errors?: Record<string, string[]>;
}

export interface BidFormProps {
  job: { id: number; title: string; budget: number };
  bid: Bid;
  errors: Record<string, string[]>;
}

File: app/models/bid.rb
# frozen_string_literal: true

class Bid < ApplicationRecord
  belongs_to :job
  belongs_to :supplier_profile

  enum :status, {
    pending: 0,
    accepted: 1,
    rejected: 2,
    withdrawn: 3
  }, default: :pending, validate: true

  validates :amount, presence: true, numericality: {greater_than: 0}
  validates :message, length: {maximum: 1000}
  validates :status, presence: true
  validates :job_id, presence: true
  validates :supplier_profile_id, presence: true, uniqueness: {scope: :job_id, message: "already placed a bid on this job"}
end
